core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -195,34 +207,69 @@ public void beforeListVariableChanged(ListVariableDescriptor<Solution_> variable
-            for (ListVariableListenerNotifiable<Solution_> notifiable : notifiables) {
+            for (var notifiable : notifiables) {
                 notifiable.notifyAfter(notification);
             }
             notificationQueuesAreEmpty = false;
         }
+        listVariableEventList.add(new ListVariableEvent(entity, fromIndex, toIndex));
     }
 
     public void triggerVariableListenersInNotificationQueues() {
-        for (Notifiable notifiable : notifiableRegistry.getAll()) {
+        for (var notifiable : notifiableRegistry.getAll()) {
             notifiable.triggerAllNotifications();
         }
+        if (listVariableDescriptor != null && !cascadingUpdateShadowVarDescriptorList.isEmpty()) {
+            triggerCascadingUpdateShadowVariableUpdate();
+        }
         notificationQueuesAreEmpty = true;
+        listVariableEventList.clear();
+    }
+
+    /**
+     * Triggers all cascading update shadow variable user-logic.
+     */
+    private void triggerCascadingUpdateShadowVariableUpdate() {
+        if (listVariableEventList.isEmpty() || cascadingUpdateShadowVarDescriptorList.isEmpty()) {
+            return;
+        }
+        for (var cascadingUpdateShadowVariableDescriptor : cascadingUpdateShadowVarDescriptorList) {
+            for (var event : listVariableEventList) {
+                var values = listVariableDescriptor.getValue(event.entity());
+                // Evaluate all elements inside the range
+                evaluateFromIndex(values, event.fromIndex(), event.toIndex(), true, cascadingUpdateShadowVariableDescriptor);
+                // Evaluate later elements, but stops when there is no change
+                evaluateFromIndex(values, event.toIndex(), values.size(), false, cascadingUpdateShadowVariableDescriptor);
+            }
+        }
+    }
+
+    private void evaluateFromIndex(List<Object> values, int fromIndex, int toIndex, boolean forceUpdate,
+            CascadingUpdateShadowVariableDescriptor<Solution_> cascadingUpdateShadowVariableDescriptor) {
+        var lastUpdated = fromIndex;
+        while (lastUpdated < toIndex) {
+            if (!cascadingUpdateShadowVariableDescriptor.update(scoreDirector, values.get(lastUpdated))
+                    && !forceUpdate) {
+                break;
+            }
+            lastUpdated++;
+        }
     }
 
     /**
      * @return null if there are no violations
      */
     public String createShadowVariablesViolationMessage() {
-        Solution_ workingSolution = scoreDirector.getWorkingSolution();
-        ShadowVariablesAssert snapshot =
+        var workingSolution = scoreDirector.getWorkingSolution();
+        var snapshot =
                 ShadowVariablesAssert.takeSnapshot(scoreDirector.getSolutionDescriptor(), workingSolution);
 
         forceTriggerAllVariableListeners(workingSolution);
 
-        final int SHADOW_VARIABLE_VIOLATION_DISPLAY_LIMIT = 3;
+        final var SHADOW_VARIABLE_VIOLATION_DISPLAY_LIMIT = 3;
```

Comment:
We don't use `final` for local variables.

Also, arguably this should be standard `camelCase`, unless you want to make it a class-level constant. (Which it probably should be.)